test/extended/images/helper.go

```diff
@@ -56,3 +80,277 @@ func CheckPageContains(oc *exutil.CLI, endpoint, path, contents string) (bool, e
+		if err != nil {
+			return err
+		}
+	}
+
+	tempDir, err := ioutil.TempDir("", "name-build")
+	if err != nil {
+		return err
+	}
+
+	dataSize := calculateRoughDataSize(oc.Stdout(), size, numberOfLayers)
+
+	lines := make([]string, numberOfLayers+1)
+	lines[0] = "FROM scratch"
+	for i := 1; i <= numberOfLayers; i++ {
+		blobName := fmt.Sprintf("data%d", i)
+		if err = createRandomBlob(path.Join(tempDir, blobName), dataSize); err != nil {
+			return err
+		}
+		lines[i] = fmt.Sprintf("COPY %s /%s", blobName, blobName)
+	}
+	if err = ioutil.WriteFile(path.Join(tempDir, "Dockerfile"), []byte(strings.Join(lines, "\n")+"\n"), 0644); err != nil {
+		return err
+	}
+
+	err = oc.Run("start-build").Args(name, "--from-dir", tempDir, "--wait").Execute()
+	out, logsErr := oc.Run("logs").Args("bc/" + name).Output()
+	if match := reSuccessfulBuild.FindStringSubmatch(out); len(match) > 1 {
+		defer dClient.RemoveImageExtended(match[1], dockerclient.RemoveImageOptions{Force: true})
+	}
+
+	if shouldSucceed && err != nil {
+		return fmt.Errorf("Got unexpected build error: %v", err)
+	}
+
+	if !shouldSucceed {
+		if err == nil {
+			return fmt.Errorf("Build unexpectedly succeeded")
+		}
+		if logsErr != nil {
+			return fmt.Errorf("Failed to show log of build config %s: %v", name, err)
+		}
+		if !reExpectedDeniedError.MatchString(out) {
+			return fmt.Errorf("Failed to match expected %q in: %q", reExpectedDeniedError.String(), out)
+		}
+	}
+
+	return nil
+}
+
+// BuildAndPushImageOfSizeWithDocker tries to build an image of wanted size and number of layers. It instructs
+// Docker daemon directly. Built image is stored as an image stream tag <name>:<tag>. If shouldSucceed is
+// false, a push is expected to fail with a denied error. Note the size is only approximate. Resulting image
+// size will be different depending on used compression algorithm and metadata overhead.
+func BuildAndPushImageOfSizeWithDocker(
+	oc *exutil.CLI,
+	dClient *dockerclient.Client,
+	name, tag string,
+	size uint64,
+	numberOfLayers int,
+	outSink io.Writer,
+	shouldSucceed bool,
+) (imageDigest string, err error) {
+	registryURL, err := GetDockerRegistryURL(oc)
+	if err != nil {
+		return "", err
+	}
+	tempDir, err := ioutil.TempDir("", "name-build")
+	if err != nil {
+		return "", err
+	}
+
+	dataSize := calculateRoughDataSize(oc.Stdout(), size, numberOfLayers)
+
+	lines := make([]string, numberOfLayers+1)
+	lines[0] = "FROM scratch"
+	for i := 1; i <= numberOfLayers; i++ {
+		blobName := fmt.Sprintf("data%d", i)
+		if err = createRandomBlob(path.Join(tempDir, blobName), dataSize); err != nil {
+			return "", err
+		}
+		lines[i] = fmt.Sprintf("COPY %s /%s", blobName, blobName)
+	}
+	if err = ioutil.WriteFile(path.Join(tempDir, "Dockerfile"), []byte(strings.Join(lines, "\n")+"\n"), 0644); err != nil {
+		return "", err
+	}
+
+	imageName := fmt.Sprintf("%s/%s/%s", registryURL, oc.Namespace(), name)
+	taggedName := fmt.Sprintf("%s:%s", imageName, tag)
```

Comment:
I dare to object. Even though the result is the same, I'm constructing neither a name for ImageStreamTag nor a reference 
to it.
On 05/26/2016 11:07 AM, Maciej Szulik wrote:
>
> In test/extended/images/helper.go <https://github.com/openshift/origin/pull/8195#discussion_r64713457>:
>
> > +
> > +	lines := make([]string, numberOfLayers+1)
> > +	lines[0] = "FROM scratch"
> > +	for i := 1; i <= numberOfLayers; i++ {
> > +		blobName := fmt.Sprintf("data%d", i)
> > +		if err = createRandomBlob(path.Join(tempDir, blobName), dataSize); err != nil {
> > +			return "", err
> > +		}
> > +		lines[i] = fmt.Sprintf("COPY %s /%s", blobName, blobName)
> > +	}
> > +	if err = ioutil.WriteFile(path.Join(tempDir, "Dockerfile"), []byte(strings.Join(lines, "\n")+"\n"), 0644); err != nil {
> > +		return "", err
> > +	}
> > +
> > +	imageName := fmt.Sprintf("%s/%s/%s", registryURL, oc.Namespace(), name)
> > +	taggedName := fmt.Sprintf("%s:%s", imageName, tag)
>
> Use |api.JoinImageStreamTag(name, tag)|
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub 
> <https://github.com/openshift/origin/pull/8195/files/0f18bf53a94fa198a09aa1504510eb66dee28a0d..2cbc04f597c0bc9374d95a8d6e718c8b31e9a7a0#r64713457>
>